### DISCUSSION TITLE
Using OR operator on columns of joined tables

### DISCUSSION THREAD

**Discussion by PAbernet** [PostgREST/postgrest#2670] [Q&A] [answered]
### Environment

* PostgreSQL version: docker.io/bitnami/postgresql:11.14.0-debian-10-r17
* PostgREST version: postgrest/postgrest:v10.1.1
* Operating system: Docker

### Description of issue
Including a column of a joined table in an OR operator results in PostgREST parsing the column/table as operator, due to it strictly adhering to the schema of `column.operator.parameters`, when it would need to be `table.column.operator.parameters`. It could really be an arbitrary number of tables preceding the filtered column.

```sql
CREATE TABLE table1 (
  id SERIAL PRIMARY KEY,
  value1 INT,
  value2 INT
);

CREATE TABLE table2 (
  id SERIAL PRIMARY KEY,
  value1 INT,
  value2 INT
);

CREATE TABLE table3 (
  id SERIAL PRIMARY KEY,
  table1_id integer,
  table2_id integer,
  FOREIGN KEY (table1_id) REFERENCES table1 (id),
  FOREIGN KEY (Table2_id) REFERENCES table2 (id)
);
```
The SQL query I'm trying to replicate:
```sql
SELECT id 
FROM table3 
  JOIN table1 ON table1_id=table1.id 
  JOIN table2 ON table2_id=table2.id 
WHERE table1.value1 > 1 OR table2.value2 > 5;
```
```
curl 'http://localhost/table3' -G -d or='(table1.value1.gt.1,table2.value2.gt.5)'
```
Is there a way to get the desired behaviour, short of creating a view?

**Comment by laurenceisla** [comment]
If I'm understanding correctly, you want to filter the result by the columns from the joined table, right? PostgREST allows this with [ebedded filters](https://postgrest.org/en/stable/api.html#embedded-filters). For your case it should be something like this:

```bash
curl 'http://localhost:3000/table2?select=*,table1(*)&table1.or=(value1.gt.1,value2.gt.5)'
```

If you want to filter the [top level rows](https://postgrest.org/en/stable/api.html#embedding-with-top-level-filtering) you need to add `!inner` to the embedding:

```bash
curl 'http://localhost:3000/table2?select=*,table1!inner(*)&table1.or=(value1.gt.1,value2.gt.5)'
```

If this isn't what you meant, then add an SQL example of what you need to check it out.

**Reply by PAbernet** [reply]
Embedded filters go in the right direction, but I need the `OR` between conditions on two different tables, which I think precludes them. I wasn't aware of them, so I updated the SQL example to reflect the more accurate requirement.

**Reply by steve-chavez** [reply]
`OR` on different tables works like described here https://github.com/PostgREST/postgrest/pull/2584

> Allows using or across related tables conditions(https://github.com/PostgREST/postgrest/discussions/2014#discussion-3670338)
curl -G 'localhost:3000/client' \
-d "select=*,clientinfo(),contact()" \
-d "clientinfo.other=ilike.*main*&contact.name=ilike.*tabby*" \
-d "or=(clientinfo.not.is.null,contact.not.is.null)"

**Reply by PAbernet** [reply]
That only seems to work on entire tables, though. Not on columns of joined tables.
```console
$ curl 'http://localhost:3000/table3' -G -d select='*,table1()' -d or='(table1.value1.gt.1,table2.value2.
gt.5)'
{"code":"PGRST100","details":"unknown single value operator value1","hint":null,"message":"\"failed to parse logic tree ((table1.value1.gt.1,table2.value2.gt.5))\" (line 1, column 18)"}
```

**Reply by PAbernet** [reply]
I feel like combining different tables in a `or` or `and` would work if the parts were indexed from the end. Without having looked at the code, I assume the parsing currently works something like this:
```python
tokens = get_param.split('.')
column = tokens[0]  # table1
operator = tokens[1]  # value1
value = tokens[2]  # gt
```
Indexing from the end could work like this:
```python
tokens = get_param.split('.')
value = tokens[-1]  # 1
operator = tokens[-2]  # gt
reference = '.'.join(tokens[,-3])  # table1.value1
```

**Reply by wolfgangwalther** [reply, marked as answer]
> That only seems to work on entire tables, though. Not on columns of joined tables.
> 
> ```
> $ curl 'http://localhost:3000/table3' -G -d select='*,table1()' -d or='(table1.value1.gt.1,table2.value2.
> gt.5)'
> {"code":"PGRST100","details":"unknown single value operator value1","hint":null,"message":"\"failed to parse logic tree ((table1.value1.gt.1,table2.value2.gt.5))\" (line 1, column 18)"}
> ```

This doesn't make sense.

The question is on which level you want to filter rows out? Do you want to filter child-level rows out, i.e. in the embedding? Then an `or` condition across multiple tables doesn't make sense.

Do you want to filter top-level rows out? Then use what @steve-chavez suggested above: Make the embedded table contain only "good" rows and then throw the whole top-level row out by checking that "whole table" against null. This also works for two tables with or.

I have a feeling you actually want something else: You want to have the "full, unfiltered embedding" **and** "filtering the top-level rows with an embedded filter". To achieve this, you need to embed the same table **twice**: Once to get all the rows in the response - and once to filter on via table filter with the reduced set of rows.
